File: packages/session-replay-react-native/ios/SRDefaultMaskingPrimitive.swift
import UIKit
import AmplitudeSessionReplay

/// Default `SRMaskingPrimitive` bridging the masking seam to the Amplitude
/// Session Replay iOS SDK's existing hook (`UIView.amp_isBlocked`):
///
///  - `maskView(_:level:)` — both `"mask"` and `"block"` map to
///    `amp_isBlocked = true` (the iOS SDK exposes a single blocking hook).
///  - `unmaskView(_:)` — `amp_isBlocked = false`.
///  - `resetView(_:)` — `amp_isBlocked = false`. Approximation: the SDK has no
///    per-view "inherit" reset today; a precise `.none` reset is a documented
///    fast-follow.
///
/// Registered on the main thread at SDK init (see `NativeSessionReplay.setup`);
/// registration replays any masking intents recorded before init (R8), so
/// mount-before-init masking still applies.
///
/// The explicit ObjC name lets the native canaries reach this class via
/// `NSClassFromString` without new public headers.
@objc(SRDefaultMaskingPrimitive)
class SRDefaultMaskingPrimitive: NSObject, SRMaskingPrimitive {
    func maskView(_ view: UIView, level: String) {
        // "mask", "block", and unknown levels all fail safe to blocking.
        view.amp_isBlocked = true
    }

    func unmaskView(_ view: UIView) {
        view.amp_isBlocked = false
    }

    func resetView(_ view: UIView) {
        view.amp_isBlocked = false
    }
}

File: packages/session-replay-react-native/android/src/main/java/com/amplitude/sessionreplayreactnative/SRDefaultMaskingPrimitive.kt
package com.amplitude.sessionreplayreactnative

import android.view.View
import com.amplitude.android.sessionreplay.SessionReplay

/**
 * Default [SRMaskingPrimitive] bridging the masking seam to the Amplitude
 * Session Replay Android SDK's existing tag-based hooks:
 *
 *  - `mask(view, "mask")`  -> [SessionReplay.mask]
 *  - `mask(view, "block")` -> [SessionReplay.block]
 *  - `unmask(view)`        -> [SessionReplay.unmask]
 *  - `reset(view)`         -> `view.tag = null` (the SDK hooks are tag-based;
 *                             clearing the tag returns the view to "inherit")
 *
 * Registered on the UI thread at SDK init
 * ([SessionReplayReactNativeModule.setup]); registration replays intents
 * recorded before init, so mount-before-init masking still applies.
 */
class SRDefaultMaskingPrimitive : SRMaskingPrimitive {
  override fun mask(view: View, level: String) {
    when (level) {
      "block" -> SessionReplay.block(view)
      // Default mask level is "mask"; unknown levels fail safe to masking.
      else -> SessionReplay.mask(view)
    }
  }

  override fun unmask(view: View) {
    SessionReplay.unmask(view)
  }

  override fun reset(view: View) {
    view.tag = null
  }
}
